openwisp_firmware_upgrader/tests/test_models.py

```diff
@@ -249,6 +249,14 @@ def test_create_for_device_validation_error(self):
         result = DeviceFirmware.create_for_device(device_fw.device)
         self.assertIsNone(result)
 
+    def test_upgrade_operation_retention_on_image_delete(self):
+        device_fw = self._create_device_firmware()
+        uo = UpgradeOperation.objects.create(
+            device=device_fw.device, image=device_fw.image
+        )
+        FirmwareImage.objects.get(pk=device_fw.image.pk).delete()
+        self.assertEqual(UpgradeOperation.objects.get(pk=uo.pk).image, None)
```

Comment:
The goal of this test is to check whether objects of `UpgradeOperation` are available even after the corresponding `FirmwareImage` is deleted. 

The current assertion makes sure such object exists since `UpgradeOperation.objects.get(pk=uo.pk)` is not raising `DoesNotExist` exception. But, it is not evident from the code at first glance. 

I suggest to first ensure that `UpgradeOperation` object exists in a assertion and then check it's image attribute is set to `None` in a separate assertion.

The above suggestion will make code more readable.